File: admin_desktop/lib/core/entities/insumo.dart
class Insumo {
  final String id;
  final String nombre;
  final String unidad;
  final double stockActual;
  final double stockMinimo;
  final double costoUnitario;
  final bool activo;

  const Insumo({
    required this.id,
    required this.nombre,
    required this.unidad,
    required this.stockActual,
    required this.stockMinimo,
    required this.costoUnitario,
    this.activo = true,
  });

  factory Insumo.fromJson(Map<String, dynamic> json) {
    return Insumo(
      id: json['id'] ?? '',
      nombre: json['nombre'] ?? '',
      unidad: json['unidad'] ?? '',
      stockActual: (json['stockActual'] as num).toDouble(),
      stockMinimo: (json['stockMinimo'] as num).toDouble(),
      costoUnitario: (json['costoUnitario'] as num).toDouble(),
      activo: json['activo'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nombre': nombre,
      'unidad': unidad,
      'stockActual': stockActual,
      'stockMinimo': stockMinimo,
      'costoUnitario': costoUnitario,
      'activo': activo,
    };
  }
}

File: functions/src/domain/entities/insumo.ts
export class Insumo {
    constructor(
        public readonly id: string,
        public nombre: string,
        public unidad: string, // 'kg', 'litro', 'pieza'
        public stockActual: number,
        public stockMinimo: number,
        public costoUnitario: number,
        public activo: boolean = true
    ) { }

    static create(props: {
        id: string;
        nombre: string;
        unidad: string;
        stockActual: number;
        stockMinimo: number;
        costoUnitario: number;
    }): Insumo {
        if (props.stockActual < 0) throw new Error("Stock cannot be negative");
        return new Insumo(
            props.id,
            props.nombre,
            props.unidad,
            props.stockActual,
            props.stockMinimo,
            props.costoUnitario
        );
    }

    actualizarStock(cantidad: number): void {
        const nuevoStock = this.stockActual + cantidad;
        if (nuevoStock < 0) throw new Error("Resulting stock cannot be negative");
        this.stockActual = nuevoStock;
    }
}
